src/test/scala/cromwell/binding/expression/TypeEvaluatorSpec.scala

```diff
@@ -0,0 +1,405 @@
+    (WdlIntegerType, "%", WdlFloatType, WdlFloatType),
+    (WdlIntegerType, "==", WdlIntegerType, WdlBooleanType),
+    (WdlIntegerType, "==", WdlFloatType, WdlBooleanType),
+    (WdlIntegerType, "!=", WdlIntegerType, WdlBooleanType),
+    (WdlIntegerType, "!=", WdlFloatType, WdlBooleanType),
+    (WdlIntegerType, "<", WdlIntegerType, WdlBooleanType),
+    (WdlIntegerType, "<", WdlFloatType, WdlBooleanType),
+    (WdlIntegerType, "<=", WdlIntegerType, WdlBooleanType),
+    (WdlIntegerType, "<=", WdlFloatType, WdlBooleanType),
+    (WdlIntegerType, ">", WdlIntegerType, WdlBooleanType),
+    (WdlIntegerType, ">", WdlFloatType, WdlBooleanType),
+    (WdlIntegerType, ">=", WdlIntegerType, WdlBooleanType),
+    (WdlIntegerType, ">=", WdlFloatType, WdlBooleanType),
+    (WdlFloatType, "+", WdlIntegerType, WdlFloatType),
+    (WdlFloatType, "+", WdlFloatType, WdlFloatType),
+    (WdlFloatType, "+", WdlStringType, WdlStringType),
+    (WdlFloatType, "-", WdlIntegerType, WdlFloatType),
+    (WdlFloatType, "-", WdlFloatType, WdlFloatType),
+    (WdlFloatType, "*", WdlIntegerType, WdlFloatType),
+    (WdlFloatType, "*", WdlFloatType, WdlFloatType),
+    (WdlFloatType, "/", WdlIntegerType, WdlFloatType),
+    (WdlFloatType, "/", WdlFloatType, WdlFloatType),
+    (WdlFloatType, "%", WdlIntegerType, WdlFloatType),
+    (WdlFloatType, "%", WdlFloatType, WdlFloatType),
+    (WdlFloatType, "==", WdlIntegerType, WdlBooleanType),
+    (WdlFloatType, "==", WdlFloatType, WdlBooleanType),
+    (WdlFloatType, "!=", WdlIntegerType, WdlBooleanType),
+    (WdlFloatType, "!=", WdlFloatType, WdlBooleanType),
+    (WdlFloatType, "<", WdlIntegerType, WdlBooleanType),
+    (WdlFloatType, "<", WdlFloatType, WdlBooleanType),
+    (WdlFloatType, "<=", WdlIntegerType, WdlBooleanType),
+    (WdlFloatType, "<=", WdlFloatType, WdlBooleanType),
+    (WdlFloatType, ">", WdlIntegerType, WdlBooleanType),
+    (WdlFloatType, ">", WdlFloatType, WdlBooleanType),
+    (WdlFloatType, ">=", WdlIntegerType, WdlBooleanType),
+    (WdlFloatType, ">=", WdlFloatType, WdlBooleanType),
+    (WdlStringType, "+", WdlIntegerType, WdlStringType),
+    (WdlStringType, "+", WdlFloatType, WdlStringType),
+    (WdlStringType, "+", WdlStringType, WdlStringType),
+    (WdlStringType, "==", WdlStringType, WdlBooleanType),
+    (WdlStringType, "!=", WdlStringType, WdlBooleanType),
+    (WdlStringType, "<", WdlStringType, WdlBooleanType),
+    (WdlStringType, "<=", WdlStringType, WdlBooleanType),
+    (WdlStringType, ">", WdlStringType, WdlBooleanType),
+    (WdlStringType, ">=", WdlStringType, WdlBooleanType),
+    (WdlFileType, "+", WdlStringType, WdlFileType),
+    (WdlFileType, "+", WdlFileType, WdlFileType),
+    (WdlFileType, "==", WdlStringType, WdlBooleanType),
+    (WdlFileType, "==", WdlFileType, WdlBooleanType),
+    (WdlFileType, "!=", WdlStringType, WdlBooleanType),
+    (WdlFileType, "!=", WdlFileType, WdlBooleanType),
+    (WdlFileType, "<=", WdlStringType, WdlBooleanType),
```

Comment:
The semantics were chosen this way because somewhat early on we decided that String and File types had "string-like" semantics in expressions.  This way you could do something like `File x = dir + "file.txt"`